### PR TITLE
[docs] Add new process to generate query help for help site

### PR DESCRIPTION
**Pull request by jf205**
This PR adds a process for generating query help documentation to use in on the CodeQL docs site so that we can move away from the Wiki.

For this new process I've added:

- a [Python script](https://github.com/github/semmle-docs/issues/272) that renders the query help as markdown. It does the following:
   - runs `codeql generate query-help` for the queries in a selected query suite for all the languages we currently support to produce a markdown file for each query help article
   -  inserts metadata, a link to the query source file and information about the query suites the query is selected by into each article
   - create `toc-<lang>.rst` files which list the query help files for each language. These toc files are part the Sphinx project to render the query help as HTML.
   
- files needed to add a new Sphinx project. This includes:
  -  a [configuration file](https://github.com/github/codeql/pull/4550/files#diff-4b5e922bee4ff8acd1cd9745b4d0292a474082588f1e2c48a5ef9579be09e8a1). It's similar to the `conf.py` files used in our other Sphinx docs, but it specifies [recommonmark](https://pypi.org/project/recommonmark/) as  an extension to build HTML from markdown.
  - a [landing page](https://github.com/github/codeql/pull/4550/files#diff-92788dc7ff4f9e37a027f27c6319a90b67be1b27a93a58cd58d4357c62ca1972) for the project
  - [a landing page](https://github.com/github/codeql/pull/4550/files#diff-0e43b01ec6288e3eb089882f5bd512d01adc5a47783d139c8efe891f689691df) for the query help for each language. These landing pages use `includes` to insert the TOCs generated by the docs generation script.

 - an [actions workflow file](https://github.com/github/codeql/pull/4550/files#diff-eb3f1a0ba13689ed77675fe29c2cc0e68e52e9f90b56c4573c7038703cb762fe) which:
   - checks out the CodeQL repos
   - obtains and sets up the CodeQL CLI
   - runs the docs generation script
   - uses [a Sphinx action](https://github.com/ammaraskar/sphinx-action) to convert the markdown file to HTML. 
   - uploads the HTML artifacts to the actions workspace. We can then download the artifact for use on the site.
 

You can see a preview of the HTML output [here](http://docteam.internal.semmle.com/james/query-help-html/).

### CODE REVIEW

**Comment by jf205** [comment]
When I ran this workflow with just the "code-scanning" and "security-extended" it took about 12 minutes. I added "security-and-quality" in https://github.com/github/codeql/pull/4550/commits/74d274c1ab7bece11388b39dfd7e4bc464195034 and now it's taking over 30 minutes 🤦

**Comment by jf205** [comment]
@sj, @esbena: the markdown files generated by the script now have filenames based on the `id` property. The short language code (e.g. `cpp`, `cs` etc) isn't included in the name because the files are already sorted into language-specific directories. See [this preview](http://docteam.internal.semmle.com/james/query-help-html/) for examples of the URLs for each query help page. Please let me know if this works for you 😸

**Comment by esbena** [comment]
Thanks. I think using `id` for file names is the right choice, but I think you should keep the short language code as well. There is nothing elsewhere that prevents queries from diverging from the convention of having a short language code as a prefix of the id.

**Comment by jf205** [comment]
> Thanks. I think using `id` for file names is the right choice, but I think you should keep the short language code as well. There is nothing elsewhere that prevents queries from diverging from the convention of having a short language code as a prefix of the id.

Ok, done 👍 

For each query help page, I've also added a line at the bottom of the query metadata section that notes which code scanning query suite the query appears in.

**Comment by jf205** [comment]
We are nearing the end of the content work for the new docs site, so I'd love to move this along if possible. It's not super important to get the processes merged, but the output need to be finalized before we use it on the new site.

With that in mind I'd like to pull in a review from @sj to make sure that we are on track from a product point of view please.

File: docs/codeql/query-help/index.rst
Comment (shati-patel) [inline docs/codeql/query-help/index.rst:14]:
```suggestion
   :caption: Table of Contents
   :titlesonly:
```


The introduction currently flows over into the TOC without a gap. Maybe add some kind of caption or header, to separate the TOC from the initial summary?

----

image

File: docs/codeql/query-help/index.rst
Comment (shati-patel) [inline docs/codeql/query-help/index.rst:10]:
```suggestion
- A link to the query in the `CodeQL repository <https://github.com/github/codeql>`__.
```

... or technically the CodeQL for Go repository. But I don't know if we need to mention that separately.

File: docs/codeql/query-help/readme.txt
Comment (shati-patel) [inline docs/codeql/query-help/readme.txt:1]: Minor: Can this README be a `.md` instead of `.txt` file? Just to get the nice formatting 🙂

File: docs/codeql/reusables/query-help-overview.rst
Comment (shati-patel) [inline docs/codeql/reusables/query-help-overview.rst:5]:
```suggestion
- ``security-and-quality``: queries from ``code-scanning``, ``security-extended``, plus extra maintainability and reliability queries.
```

File: docs/codeql/query-help/go.rst
Comment (shati-patel) [inline docs/codeql/query-help/go.rst:6]:
```suggestion
For shorter queries that you can use as building blocks when writing your own queries, see the `example queries in the CodeQL for Go repository <https://github.com/github/codeql-go/tree/main/ql/examples>`__.
```

Again, I don't know if it's necessary to make this distinction 🤷🏽

File: docs/codeql/conf.py
Comment (shati-patel) [inline docs/codeql/conf.py:113]:
```suggestion
##############################################################################

```

Add a newline at the end of the file, just to keep the UNIX gods happy (or something like that) 🙈

File: docs/codeql/query-help/conf.py
Comment (shati-patel) [inline docs/codeql/query-help/conf.py:60]:
```suggestion
exclude_patterns = ['toc-*'] # ignore toc-<lang>.rst files as they are 'included' in index pages

```
Another newline at the end of the file

File: docs/codeql/query-help/conf.py
Comment (shati-patel) [inline docs/codeql/query-help/conf.py:33]:
```suggestion
# -- Project-specific options for HTML output ----------------------------------------------
```

I'm aware this doesn't matter, but hey 😉

File: docs/codeql/query-help/index.rst
Comment (shati-patel) [inline docs/codeql/query-help/index.rst:11]:
```suggestion
- A description of the potential vulnerability that the query identifies and a recommendation for how to avoid introducing the problem to your code.
```

Minor tweak

File: docs/codeql/reusables/query-help-overview.rst
Comment (shati-patel) [inline docs/codeql/reusables/query-help-overview.rst:3]:
```suggestion
- ``code-scanning``: queries run by default in CodeQL code scanning on GitHub.
```

Plus maybe a link to [GitHub docs](https://docs.github.com/github/finding-security-vulnerabilities-and-errors-in-your-code/about-code-scanning)? 


Perhaps not necessary at all, but it could help people who aren't familiar with the "code scanning" feature on GitHub.

**Review by shati-patel — COMMENTED** [review]
Yay, this looks great @jf205! 🎉 I've added a couple of minor (mostly optional) suggestions and questions. 

I haven't looked too closely at the script/workflow, but let me know if there's anything specific you'd like me to test or review!

File: docs/codeql/query-help/index.rst
Comment (jf205) [reply docs/codeql/query-help/index.rst:14]: Ah yes, this does look odd. I don't think showing the full (expanded) TOC is very helpful either, so I'll fix that.

File: docs/codeql/query-help/readme.txt
Comment (jf205) [reply docs/codeql/query-help/readme.txt:1]: 👍🏻

Comment (shati-patel) [reply docs/codeql/query-help/readme.txt]: Thanks! And good point about excluding it from the build - I hadn't thought about that 👏🏽

**Review by shati-patel — APPROVED** [review]
Thanks for the tweaks! All LGTM from a docs point of view, but I'll let someone more qualified take a look at the scripts 👀

**Comment by adityasharad** [comment]
Thanks! I'll do a proper review this week. I think some of the functionality you've written should eventually make its way into `codeql generate query-help`, and then we can run it internally against all the relevant repos, but that doesn't need to hold up this PR.

**Comment by jf205** [comment]
> Thanks! I'll do a proper review this week. I think some of the functionality you've written should eventually make its way into codeql generate query-help, and then we can run it internally against all the relevant repos, but that doesn't need to hold up this PR.

👍🏻 thanks @adityasharad 

Merging this now 🎉

**Comment by adityasharad** [comment]
Quick follow-up PR: https://github.com/github/codeql/pull/4770

**Pull request merged**
